### ISSUE TITLE
Scheduler crash

### ISSUE THREAD

**Issue by SeMeKh** [kubernetes/kubernetes#38105]
**Is this a BUG REPORT or FEATURE REQUEST?** (choose one): Bug Report

**Kubernetes version** (use `kubectl version`):

```
Client Version: version.Info{Major:"1", Minor:"4", GitVersion:"v1.4.0", GitCommit:"a16c0a7f71a6f93c7e0f222d961f4675cd97a46b", GitTreeState:"clean", BuildDate:"2016-09-26T18:16:57Z", GoVersion:"go1.6.3", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"4", GitVersion:"v1.4.0", GitCommit:"a16c0a7f71a6f93c7e0f222d961f4675cd97a46b", GitTreeState:"clean", BuildDate:"2016-09-26T18:10:32Z", GoVersion:"go1.6.3", Compiler:"gc", Platform:"linux/amd64"}
```

**Environment**:
- **Cloud provider or hardware configuration**: N/A
- **OS** (e.g. from /etc/os-release): `Ubuntu 16.04 LTS (Xenial Xerus)`
- **Kernel** (e.g. `uname -a`): `Linux 4.4.0-47-generic #68-Ubuntu SMP Wed Oct 26 19:39:52 UTC 2016 x86_64 x86_64 x86_64 GNU/Linux`

**What happened**: A node was out of disk, and a Pod from a PetSet (with `replicas=3`) was deleted but it was not rescheduled. The scheduler kept crashing instead. Deleting the node marked as OutOfDisk fixed the crashes.

**How to reproduce it** (as minimally and precisely as possible): Not easily reproducible now that it's gone. Happened in our production servers only. 

**Anything else do we need to know**:

Logs from last kubernetes scheduler crash:
```
I1205 11:34:05.631075       1 factory.go:257] Creating scheduler from algorithm provider 'DefaultProvider'
I1205 11:34:05.633047       1 factory.go:303] creating scheduler with fit predicates 'map[NoVolumeZoneConflict:{} GeneralPredicates:{} PodToleratesNodeTaints:{} NoDiskConflict:{} MaxEBSVolumeCount:{} MaxGCEPDVolumeCount:{} CheckNodeMemoryPressure:{} CheckNodeDiskPressure:{} MatchInterPodAffinity:{}]' and priority functions 'map[TaintTolerationPriority:{} InterPodAffinityPriority:{} LeastRequestedPriority:{} BalancedResourceAllocation:{} SelectorSpreadPriority:{} NodePreferAvoidPodsPriority:{} NodeAffinityPriority:{}]
I1205 11:34:05.633265       1 reflector.go:211] Starting reflector *api.Pod (0) from k8s.io/kubernetes/plugin/pkg/scheduler/factory/factory.go:389
I1205 11:34:05.633324       1 reflector.go:211] Starting reflector *api.PersistentVolume (0) from k8s.io/kubernetes/plugin/pkg/scheduler/factory/factory.go:399
I1205 11:34:05.633367       1 reflector.go:211] Starting reflector *api.PersistentVolumeClaim (0) from k8s.io/kubernetes/plugin/pkg/scheduler/factory/factory.go:400
I1205 11:34:05.633393       1 reflector.go:211] Starting reflector *api.Service (0) from k8s.io/kubernetes/plugin/pkg/scheduler/factory/factory.go:405
I1205 11:34:05.633426       1 reflector.go:211] Starting reflector *api.ReplicationController (0) from k8s.io/kubernetes/plugin/pkg/scheduler/factory/factory.go:410
I1205 11:34:05.633450       1 reflector.go:211] Starting reflector *extensions.ReplicaSet (0) from k8s.io/kubernetes/plugin/pkg/scheduler/factory/factory.go:415
I1205 11:34:05.633383       1 reflector.go:249] Listing and watching *api.Pod from k8s.io/kubernetes/plugin/pkg/scheduler/factory/factory.go:389
I1205 11:34:05.633663       1 reflector.go:211] Starting reflector *api.Pod (0) from k8s.io/kubernetes/plugin/pkg/scheduler/factory/factory.go:392
I1205 11:34:05.633704       1 reflector.go:249] Listing and watching *api.Pod from k8s.io/kubernetes/plugin/pkg/scheduler/factory/factory.go:392
I1205 11:34:05.633877       1 reflector.go:249] Listing and watching *api.Service from k8s.io/kubernetes/plugin/pkg/scheduler/factory/factory.go:405
I1205 11:34:05.634079       1 reflector.go:249] Listing and watching *api.PersistentVolume from k8s.io/kubernetes/plugin/pkg/scheduler/factory/factory.go:399
I1205 11:34:05.634356       1 reflector.go:211] Starting reflector *api.Node (0) from k8s.io/kubernetes/plugin/pkg/scheduler/factory/factory.go:395
I1205 11:34:05.634352       1 reflector.go:249] Listing and watching *api.PersistentVolumeClaim from k8s.io/kubernetes/plugin/pkg/scheduler/factory/factory.go:400
I1205 11:34:05.634408       1 reflector.go:249] Listing and watching *api.Node from k8s.io/kubernetes/plugin/pkg/scheduler/factory/factory.go:395
I1205 11:34:05.634496       1 reflector.go:249] Listing and watching *extensions.ReplicaSet from k8s.io/kubernetes/plugin/pkg/scheduler/factory/factory.go:415
I1205 11:34:05.634517       1 reflector.go:249] Listing and watching *api.ReplicationController from k8s.io/kubernetes/plugin/pkg/scheduler/factory/factory.go:410
E1205 11:34:05.677622       1 event.go:258] Could not construct reference to: '&api.Endpoints{TypeMeta:unversioned.TypeMeta{Kind:"", APIVersion:""}, ObjectMeta:api.ObjectMeta{Name:"kube-scheduler", GenerateName:"", Namespace:"kube-system", SelfLink:"", UID:"", ResourceVersion:"", Generation:0, CreationTimestamp:unversioned.Time{Time:time.Time{sec:0, nsec:0, loc:(*time.Location)(nil)}}, DeletionTimestamp:(*unversioned.Time)(nil), DeletionGracePeriodSeconds:(*int64)(nil), Labels:map[string]string(nil), Annotations:map[string]string(nil), OwnerReferences:[]api.OwnerReference(nil), Finalizers:[]string(nil), ClusterName:""}, Subsets:[]api.EndpointSubset(nil)}' due to: 'selfLink was empty, can't make reference'. Will not report event: 'Normal' '%v became leader' 'k8s-1'
I1205 11:34:05.677852       1 leaderelection.go:214] sucessfully acquired lease kube-system/kube-scheduler
I1205 11:34:05.688019       1 factory.go:422] About to try and schedule pod game-connectivity-api-1875989870-y452d
I1205 11:34:05.688108       1 scheduler.go:96] Attempting to schedule pod: backtory/game-connectivity-api-1875989870-y452d
I1205 11:34:05.688774       1 factory.go:582] Attempting to bind game-connectivity-api-1875989870-y452d to k8s-4
I1205 11:34:05.688848       1 factory.go:422] About to try and schedule pod realtime-api-0
I1205 11:34:05.688864       1 scheduler.go:96] Attempting to schedule pod: backtory/realtime-api-0
I1205 11:34:05.689357       1 factory.go:582] Attempting to bind realtime-api-0 to k8s-3
I1205 11:34:05.700591       1 event.go:217] Event(api.ObjectReference{Kind:"Pod", Namespace:"backtory", Name:"game-connectivity-api-1875989870-y452d", UID:"b9b5599a-bade-11e6-9573-00505692e5bb", APIVersion:"v1", ResourceVersion:"6614927", FieldPath:""}): type: 'Normal' reason: 'Scheduled' Successfully assigned game-connectivity-api-1875989870-y452d to k8s-4
I1205 11:34:05.708732       1 event.go:217] Event(api.ObjectReference{Kind:"Pod", Namespace:"backtory", Name:"realtime-api-0", UID:"ba077781-bade-11e6-9573-00505692e5bb", APIVersion:"v1", ResourceVersion:"6614961", FieldPath:""}): type: 'Normal' reason: 'Scheduled' Successfully assigned realtime-api-0 to k8s-3
I1205 11:34:07.685132       1 leaderelection.go:228] succesfully renewed lease kube-system/kube-scheduler
I1205 11:34:08.356372       1 factory.go:422] About to try and schedule pod realtime-api-1
I1205 11:34:08.356415       1 scheduler.go:96] Attempting to schedule pod: backtory/realtime-api-1
E1205 11:34:08.356770       1 runtime.go:64] Observed a panic: "invalid memory address or nil pointer dereference" (runtime error: invalid memory address or nil pointer dereference)
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/util/runtime/runtime.go:70
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/util/runtime/runtime.go:63
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/util/runtime/runtime.go:49
/usr/local/go/src/runtime/asm_amd64.s:472
/usr/local/go/src/runtime/panic.go:443
/usr/local/go/src/runtime/panic.go:62
/usr/local/go/src/runtime/sigpanic_unix.go:24
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/plugin/pkg/scheduler/algorithm/predicates/predicates.go:971
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/plugin/pkg/scheduler/algorithm/predicates/predicates.go:1002
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/plugin/pkg/scheduler/algorithm/predicates/predicates.go:830
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/plugin/pkg/scheduler/algorithm/predicates/predicates.go:822
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/plugin/pkg/scheduler/generic_scheduler.go:215
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/plugin/pkg/scheduler/generic_scheduler.go:167
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/util/workqueue/parallelizer.go:43
/usr/local/go/src/runtime/asm_amd64.s:1998
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal 0xb code=0x1 addr=0x10 pc=0xc97311]

goroutine 354 [running]:
panic(0x22a7ac0, 0xc82000e070)
	/usr/local/go/src/runtime/panic.go:481 +0x3e6
k8s.io/kubernetes/pkg/util/runtime.HandleCrash(0x0, 0x0, 0x0)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/util/runtime/runtime.go:56 +0x153
panic(0x22a7ac0, 0xc82000e070)
	/usr/local/go/src/runtime/panic.go:443 +0x4e9
k8s.io/kubernetes/plugin/pkg/scheduler/algorithm/predicates.(*PodAffinityChecker).getMatchingAntiAffinityTerms(0xc820256740, 0xc820380000, 0xc8204e5000, 0x54, 0x80, 0x0, 0x0, 0x0, 0x0, 0x0)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/plugin/pkg/scheduler/algorithm/predicates/predicates.go:971 +0x131
k8s.io/kubernetes/plugin/pkg/scheduler/algorithm/predicates.(*PodAffinityChecker).satisfiesExistingPodsAntiAffinity(0xc820256740, 0xc820380000, 0x0, 0x0, 0xc8201e4278, 0xc8206cd1d0)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/plugin/pkg/scheduler/algorithm/predicates/predicates.go:1002 +0x986
k8s.io/kubernetes/plugin/pkg/scheduler/algorithm/predicates.(*PodAffinityChecker).InterPodAffinityMatches(0xc820256740, 0xc820380000, 0x0, 0x0, 0xc8202a7bc0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/plugin/pkg/scheduler/algorithm/predicates/predicates.go:830 +0x117
k8s.io/kubernetes/plugin/pkg/scheduler/algorithm/predicates.(*PodAffinityChecker).InterPodAffinityMatches-fm(0xc820380000, 0x0, 0x0, 0xc8202a7bc0, 0x1, 0x0, 0x0, 0x0, 0x0, 0x0)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/plugin/pkg/scheduler/algorithm/predicates/predicates.go:822 +0x7c
k8s.io/kubernetes/plugin/pkg/scheduler.podFitsOnNode(0xc820380000, 0x0, 0x0, 0xc8202a7bc0, 0xc820079680, 0xc82049ee50, 0x0, 0x0, 0x0, 0x0, ...)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/plugin/pkg/scheduler/generic_scheduler.go:215 +0x10e
k8s.io/kubernetes/plugin/pkg/scheduler.findNodesThatFit.func1(0x1)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/plugin/pkg/scheduler/generic_scheduler.go:167 +0x169
k8s.io/kubernetes/pkg/util/workqueue.Parallelize.func1(0xc820509210, 0xc8203eb800, 0xc82038bea0)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/util/workqueue/parallelizer.go:43 +0xbd
created by k8s.io/kubernetes/pkg/util/workqueue.Parallelize
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/util/workqueue/parallelizer.go:45 +0x10f
```

**Comment by bprashanth** [comment]
I think we fixed one of these in a minor release, @kubernetes/sig-scheduling would know

**Comment by davidopp** [comment]
Yes this was fixed, I can look up the PR that fixed it later, but the fix is definitely in the latest 1.4.

**Comment by timothysc** [comment]
I think you are referring to #29090 , but timeframe and signature look different.

**Comment by timothysc** [comment]
We'll trace down if it's the same, but the timeframes don't match up.

**Comment by davidopp** [comment]
No I was thinking of #28350 / #31968 which are fixed by #33086.

So I'm going to close this, but please reopen if you think this is actually a new issue.

**Comment by jayunit100** [comment]
yup, 
```
k8s.io/kubernetes/plugin/pkg/scheduler/algorithm/predicates.(*PodAffinityChecker).getMatchingAntiAffinityTerms(0xc820256740, 0xc820380000, 0xc8204e5000, 0x54, 0x80, 0x0, 0x0, 0x0, 0x0, 0x0)
```
Is specifically what 33086 is fixing.  So good to keep closed.
